### PR TITLE
Remove \OCP\App::register()

### PR DESCRIPTION
**Pull request by MorrisJobke**
This method can be removed. This isn't used anymore since 4.5.0.

It's deprecated since mid 2012 https://github.com/owncloud/core/commit/e4679770c4d85896bef3e81125e86e272bb6cd64

It doesn't do anything since mid 2012 https://github.com/owncloud/core/commit/8a92cd21d6768fc27c4892583bef327da9a7e5f4

cc @karlitschek @DeepDiver1975 @LukasReschke @PVince81

### CODE REVIEW

**Comment by karlitschek** [comment]
Hmmm. Why remove it? This would break old apps that still call it?
We should try to keep the public API stable as long as possible. 
:-1:

**Comment by LukasReschke** [comment]
Mhm. I'd just add here that a 2012 app won't anyways run anymore properly in ownCloud and requires adjusting since we actually had other breaking changes such as the routing. Considering that I'd not really consider this as a very big deal. It won't work anyways and if devs have to fix it they can also remove this in the same step :-)

**Comment by DeepDiver1975** [comment]
That function was deprecated back in 2012 - https://github.com/owncloud/core/commit/e4679770c4d85896bef3e81125e86e272bb6cd64#diff-0b49ed7c07cd77890522003e869529c0

We did agree on to remove deprecated apis after 2 years - so removing this API is valid from my pov.

**Comment by DeepDiver1975** [comment]
Some apps still use it - https://github.com/search?utf8=%E2%9C%93&q=%22ocp+App%3A%3Aregister%22+NOT+registerAdmin+NOT+registerPersonal&type=Code&ref=searchresults

.. questionable if they work at all on 8.1 indeed

**Comment by karlitschek** [comment]
Sure. It is questionable if this apps work at all. But our goal is to keep the api stable for many years. So we should try to do that.

**Comment by DeepDiver1975** [comment]
> many years

= :two: years?

**Comment by rullzer** [comment]
So just my two cents on this....

Deprecated stuff has to go at some point. It serves a purpose. It forces app developers to use the new code. Which is what is patched an maintained. There is a 2 year deprecation period so I would argue that that is enough time to move to the new API.

Also, most people already move away from the deprecated stuff. Which means that the deprecated code gets tested less and less over time. In general we do not write new unit tests for deprecated code. Since well the function is deprecated anyway, and you might open pandoras box..

With a deprecation period of 2 years basically the policy is we have to still fix stuff. After those 2 years nobody touches it anymore (and most likely also not before). So keeping the code around would only be valid if we are absolutely certain that our unit tests cover all the possible cases so we know for sure no bad stuff happens.

Else you could be shipping broken code for a long time. If for example (extreme) an app deletes all data of a user because of some bug in a deprecated function. The user probably does not care that that that function should not have been used. 

Long story short. There is a reason functions get deprecated. And if it is decided to have a 2 year period. Then I would argue it is time to remove this function.

**Comment by karlitschek** [comment]
the goal is to keep the api longer stable then just 2 years.

**Comment by DeepDiver1975** [comment]
Deprecation naturally results in deletion after some point in time - otherwise there is no need in deprecating. Usually APIs are deprecated for a reason and we need to cleanup the code base.

2 years seems a reasonable time frame for me. :+1: 

@karlitschek Does that make sense?

**Comment by MorrisJobke** [comment]
@LukasReschke @nickvergessen Review?

**Comment by VicDeo** [comment]
@DeepDiver1975 
https://github.com/search?utf8=%E2%9C%93&q=%22ocp+App%3A%3Aregister%22+NOT+registerAdmin+NOT+registerPersonal&type=Code&ref=searchresults

It's funny to know that nearly a half of search results by this link uses appid  `impress` for this call
```
OCP\App::register(array('order' => 70, 'id' => 'impress', 'name' => 'Impress'));
```
other popular appids are 
```
'id' => 'gallery',  'name' => 'Pictures'
```
and
```
 'id' => 'user_webfinger',  'name' => 'Webfinger'
```

**Comment by VicDeo** [comment]
@MorrisJobke requires removal of https://github.com/owncloud/apps/blob/master/reader/appinfo/app.php#L4  at least

**Comment by LukasReschke** [comment]
The apps in the ~~graveyard~~ apps repo are anyways nearly all completely broken due to other changes we did.

And yes, I personally am not a big fan of this repository. - It is claiming that the apps in there are suitable for use.

**Comment by karlitschek** [comment]
then we have to fix them and not make incomoatible changes in core. thanks

**Comment by LukasReschke** [comment]
> then we have to fix them and not make incomoatible changes in core. thanks

I beg to differ here. In my opinion we should not spend time fixing third-party apps that are unchanged since 2-3 years.

Just as a reality check I checked when those apps got last touched with more than a simple "search and replace that nobody tested"  or the translations sync.
<hr/>
- **admin_migrate:** 1 year ago, won't work anymore at all since the migrate functionality was removed in core.
- **apptemplate:** 2 years ago and replaced by https://github.com/owncloud/ocdev/tree/6870aefe0937d7a7746abdb0a478fd4bb2ebc26f/ocdev/plugins/startapp/templates
- **django_auth:** 2 years ago, insecure as it misses at least CSRF checks (didn't check further)
- **emoji:** 2 years ago, and honestly, what is that app supposed to do? "A plugin for other Apps to add Emoji support. Those Apps have to support plugins" – I doubt that any app really uses this and then again this is something that a library should do and not a single app…
- **encryption_dummy:** Okay. Fair enough.
- **external:** 8 months ago - fair enough, we ship this so it actually gets looked at from time to time.
- **files_archive: **8 months ago - fair enough, we used to ship this.
- **files_odfviewer:** Replaced by the documents app
- **files_sgfviewer:** 3 years ago, has XSS bugs, probably doesn't work at all anymore though. "Review go games (.sgf files)", all right…
- **files_svgedit:** 3 years ago, not sure if this will work as well.
- **files_videoviewer:** Works somewhat.
- **fluxx_compensator:** 1 year ago. Left there to die by the maintainer, maybe works, maybe not.
- **impress:** 2 years ago. Makes ownCloud insecure by design. - Also unmaintained.
- **imprint:** Woah, the first really maintained one :see_no_evil: 
- **ownpad_lite:** 2 years ago. Won't work anymore due to other changes in the routing.
- **pong:** That one should work as it has no PHP logic.
- **pushnotifications:** Should also work. Untested.
- **reader:** 2 years ago, won't work due to other changes.
- **search:** 2 years ago, won't work due to other changes.
- **tattoo:** 2 years ago, won't work due to other changes.
- **user_external:** Maintained one by us.
- **user_migrate:** 1 year ago, won't work due to removing the API.
- **user_oauth:** 2 years ago, won't work due to other changes.
- **user_openid_provider:** 3 years ago, won't work due to other changes.
- **user_persona:** 1 year ago, won't work due to other changes.
- **user_saml:** 9 months ago ("make compatible with ownCloud 6"), won't work in newer versions due to other changes.
- **user_vd:** 2 years ago, won't work due to other changes.
- **user_webfinger:** 3 years ago, no idea what it does, code looks fishy.

<hr/>

So what we have is a huge bunch of age from the medieval age that nobody really cares about. Most of them actually don't work at the moment, and even if we fix the PHP parts they will still be broken as the JS / HTML of ownCloud changed a lot in the last 3 years.

A check here will simply show that apps that use the since years deprecated function discussed here will simply not work at all due to other changes. 3 years is a long time and we changed a lot of stuff, sometimes we can't prevent backwards compatible changes for security reasons or because we have to move on with the code base to ensure that it is somewhat maintainable.

The apps repo is a huge red herring. I'd really appreciate if we could move the apps into their own repositories, that would at least show all how unmaintained they are. Just like a bunch of other repositories that we have: https://github.com/owncloud?page=6

I find it personally rather worrisome that we have that many unmaintained apps in the owncloud organization. Users *don't* understand that if they go to "github.com/owncloud" that the code they download actually is *not* from ownCloud. I hear all the time in IRC and other channels complaints about "us having some broken apps in the repository". - All the stuff that we offer from the "owncloud" organization is in real-life expected by most users to be of us or at least to work somewhat. (as long as the repository is not markd with a huge red abandonware warning.

**Comment by LukasReschke** [comment]
That said: I agree that we should keep API stable where possible. But in some cases it is in my opinion the right thing to remove stuff, one of it is the API is unused and doing nothing since ages. 

Especially if any competent dev that cares about their app can fix their code in about 20 minutes.

**Comment by karlitschek** [comment]
i think we have here a fundamentally broken understanding what it means to maintain a stable platform and trying to build an ecosystem of independent apps and developers. we have to discuss this at the next possible oportunity

**Comment by PVince81** [comment]
More info here in a similar discussion: https://github.com/owncloud/documentation/issues/654#issuecomment-63465998

**Comment by DeepDiver1975** [comment]
> then we have to fix them and not make incomoatible changes in core. thanks

we cannot fix the whole universeof all apps out there. regarding incompatible changes:

we have to fix broken apis and sometimes this results in deprecation and removal after some time.
We will have even bigger changes then this simple function in the future: \OCP\Share - the api is broken and needs an overhaul.

let's address this all in 8.2

**Comment by MorrisJobke** [comment]
@karlitschek @DeepDiver1975 @LukasReschke @PVince81 Can we have a decision here? Either we merge this or I will close this PR and remove my change. I'm fine with both.

**Comment by DeepDiver1975** [comment]
As specified - and finally set to stone - deprecated APIs will be removed after 3 years. http://karlitschek.de/2015/06/building-a-platform/
So this one is to be dropped October this year.

8.2 will be released in October -> :+1:

**Comment by karlitschek** [comment]
agreed :+1:
